src/fork_gc.c

```diff
@@ -322,177 +344,179 @@ static void ForkGc_CollectGarbageFromTagIdx(ForkGCCtx *gc, RedisSearchCtx *sctx)
-    }
-    rm_free(idxData->freeBufs);
+static void ForkGc_FixInvertedIndex(ForkGCCtx *gc, InvIdxInfo *idxData, InvertedIndex *idx) {
+  for (size_t i = 0; i < idxData->numAddrsToFree; ++i) {
+    rm_free(idxData->addrsToFree[i]);
   }
+  rm_free(idxData->addrsToFree);
 
-  for (int i = 0; i < array_len(idxData->blocksModified); ++i) {
-    ModifiedBlock *blockModified = idxData->blocksModified + i;
-    indexBlock_Free(&idx->blocks[blockModified->blockOldIndex]);
+  if (idxData->numChangedBlocks) {
+    for (size_t i = 0; i < idxData->numChangedBlocks - 1; ++i) {
+      ModifiedBlock *blockModified = idxData->changedBlocks + i;
+      indexBlock_Free(&idx->blocks[blockModified->blockOldIndex]);
+    }
+    ModifiedBlock *lastBinfo = idxData->changedBlocks + idxData->numChangedBlocks - 1;
+    IndexBlock *lastOld = idx->blocks + lastBinfo->blockOldIndex;
+    if (lastBinfo->blk.numDocs == lastOld->numDocs) {
+      indexBlock_Free(lastOld);
+    }
   }
 
+  // Ensure the old index is at least as big as the new index' size
   assert(idx->size >= idxData->originalSize);
-  if (idxData->blockArrayExists) {
-    idxData->newBlocksArray = rm_realloc(
-        idxData->newBlocksArray,
-        (idxData->newBlocksArraySize +
-         (idx->size - (idxData->originalSize - 1 /* we are copy the last block anyway*/))) *
-            sizeof(IndexBlock));
-    memcpy(idxData->newBlocksArray + idxData->newBlocksArraySize,
-           idx->blocks + (idxData->originalSize - 1),
-           (idx->size - (idxData->originalSize - 1)) * sizeof(IndexBlock));
+  if (idxData->newBlocks) {
+    /**
+     * 'newBlocks' means reordered blocks [TODO: (Mark) rename this variable].
+     * If we have a new block ordering, let the current 'newBlocks' array be
+     * the indexes' new block array
+     */
+
+    // Number of blocks added in the parent process since the last scan
+    size_t newAddedLen = idx->size - idxData->originalSize;
+
+    // The final size is the reordered block size, plus the number of blocks
+    // which we haven't scanned yet, because they were added in the parent
+    size_t totalLen = idxData->numNewBlocks + newAddedLen;
+
+    idxData->newBlocks = rm_realloc(idxData->newBlocks, totalLen * sizeof(*idxData->newBlocks));
+    memcpy(idxData->newBlocks + idxData->numNewBlocks, (idx->blocks + idxData->originalSize),
+           newAddedLen * sizeof(*idxData->newBlocks));
     rm_free(idx->blocks);
-    idxData->newBlocksArraySize += (idx->size - (idxData->originalSize - 1));
-    idx->blocks = idxData->newBlocksArray;
-    idx->size = idxData->newBlocksArraySize;
+    idxData->numNewBlocks += newAddedLen;
+    idx->blocks = idxData->newBlocks;
+    idx->size = idxData->numNewBlocks;
   }
 
-  for (size_t i = 0; i < array_len(idxData->blocksModified); ++i) {
-    ModifiedBlock *blockModified = idxData->blocksModified + i;
+  for (size_t i = 0; i < idxData->numChangedBlocks; ++i) {
+    ModifiedBlock *blockModified = idxData->changedBlocks + i;
```

Comment:
Once again you might override the last index, or am I missing something?

Reply:
It looks like I'll need to make a new function handling the last block specifically..